### PR TITLE
feat: use faster query_and_wait method from google-cloud-bigquery when available

### LINKED ISSUE #710: use faster `query_and_wait` method when available to skip BQ Storage API if all query results are already available
Similar to https://github.com/googleapis/python-bigquery/issues/1745, pandas-gbq should benefit from the optimizations done in https://github.com/googleapis/python-bigquery/issues/589.

This might require a progress bar feature in `query_and_wait` if we want to keep the progress bar for waiting for the query to finish, but we need to be careful when introducing such a progress bar to avoid timing out API requests too soon.

### PR DESCRIPTION
**Pull request by tswast**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-pandas/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Requires #720, #721, and https://github.com/googleapis/python-bigquery/pull/1793 to be merged first.

Fixes #710 🦕

### CODE REVIEW

**Comment by tswast** [comment]
I tested this manually in a local ipython instance to verify that the latency is improved when using this new functionality:

**After** this change:

```python
# In [1]:
import pandas_gbq

# In [2]: %%timeit -n10 -r10
df = pandas_gbq.read_gbq("SELECT 1", progress_bar_type=None)

# Out [2]:
# 394 ms ± 37.8 ms per loop (mean ± std. dev. of 10 runs, 10 loops each)
```

**Before** this change:

```python
# In [1]:
import pandas_gbq

# In [2]: %%timeit -n10 -r10
df = pandas_gbq.read_gbq("SELECT 1", progress_bar_type=None)

# Out [2]:
# 1.08 s ± 102 ms per loop (mean ± std. dev. of 10 runs, 10 loops each)
```

This is quite similar to the results I observed when implementing this optimization in the DB-API connector. https://github.com/googleapis/python-bigquery/pull/1747#issuecomment-1852368885

File: tests/system/test_gbq.py
Comment (tswast) [inline tests/system/test_gbq.py:429]: I'm not 100% sure why this isn't failing anymore. I guess jobs.query is more lenient about unknown parameters than jobs.insert is?

Comment (Linchin) [reply tests/system/test_gbq.py:429]: Could you share a link to a failing test log? I wasn't able to find it in test fusion.

Comment (Linchin) [reply tests/system/test_gbq.py:429]: Should we keep this test?

**Comment by Linchin** [comment]
All looks good except for that I wanted to double check on the removed system test.

File: tests/system/test_gbq.py
Comment (tswast) [reply tests/system/test_gbq.py:429]: I tested locally with https://github.com/googleapis/python-bigquery/pull/1793 and the request body now includes unknown parameters, but surprisingly the REST API seems to ignore these. I'll follow-up on that PR to raise on common errors, such as passing in a configuration object of the wrong type.

Comment (tswast) [reply tests/system/test_gbq.py]: Updated https://github.com/googleapis/python-bigquery/pull/1793 with some client-side checks for common mistakes of sending the wrong type for `job_config` and now this test is passing.

Will need to wait for a `google-cloud-bigquery` release with that fix before merging this PR.

**Comment by tswast** [comment]
@Linchin Tests are passing after the most recent google-cloud-bigquery release. OK for another look.

**Comment by Linchin** [comment]
LGTM, thank you!

**Review by Linchin — APPROVED** [review]

**Pull request merged**
